src/reducers/mod.rs

```diff
@@ -1,79 +1,120 @@
+            Plugin::PointByTx(x) => x.reduce_block(block, output),
+            Plugin::PoolByStake(x) => x.reduce_block(block, output),
+            Plugin::TotalTransactionsCount(x) => x.reduce_block(block, output),
+            Plugin::TransactionsCountByEpoch(x) => x.reduce_block(block, output),
+            Plugin::TransactionsCountByContractAddress(x) => x.reduce_block(block, output),
+            Plugin::TransactionsCountByContractAddressByEpoch(x) => x.reduce_block(block, output),
+            Plugin::TotalTransactionsCountByContractAddresses(x) => x.reduce_block(block, output),
         }
     }
+}
 
-    pub fn borrow_output_port(&mut self) -> &'_ mut OutputPort<model::CRDTCommand> {
-        match self {
-            Plugin::UtxoByAddress(x) => x.borrow_output_port(),
-            Plugin::PointByTx(x) => x.borrow_output_port(),
-            Plugin::PoolByStake(x) => x.borrow_output_port(),
-            Plugin::TotalTransactionsCount(x) => x.borrow_output_port(),
-            Plugin::TransactionsCountByEpoch(x) => x.borrow_output_port(),
-            Plugin::TransactionsCountByContractAddress(x) => x.borrow_output_port(),
-            Plugin::TransactionsCountByContractAddressByEpoch(x) => x.borrow_output_port(),
-            Plugin::TotalTransactionsCountByContractAddresses(x) => x.borrow_output_port(),
+pub struct Worker {
+    input: InputPort,
+    output: OutputPort,
+    reducers: Vec<Plugin>,
+    ops_count: gasket::metrics::Counter,
+}
+
+impl Worker {
+    pub fn new(reducers: Vec<Plugin>) -> Self {
+        Worker {
+            reducers,
+            input: Default::default(),
+            output: Default::default(),
+            ops_count: Default::default(),
         }
     }
 
+    pub fn borrow_input_port(&mut self) -> &'_ mut InputPort {
+        &mut self.input
+    }
+
+    pub fn borrow_output_port(&mut self) -> &'_ mut OutputPort {
+        &mut self.output
+    }
+
     pub fn spawn(self, pipeline: &mut bootstrap::Pipeline) {
-        match self {
-            Plugin::UtxoByAddress(x) => x.spawn(pipeline),
-            Plugin::PointByTx(x) => x.spawn(pipeline),
-            Plugin::PoolByStake(x) => x.spawn(pipeline),
-            Plugin::TotalTransactionsCount(x) => x.spawn(pipeline),
-            Plugin::TransactionsCountByEpoch(x) => x.spawn(pipeline),
-            Plugin::TransactionsCountByContractAddress(x) => x.spawn(pipeline),
-            Plugin::TransactionsCountByContractAddressByEpoch(x) => x.spawn(pipeline),
-            Plugin::TotalTransactionsCountByContractAddresses(x) => x.spawn(pipeline),
-        }
+        pipeline.register_stage("reducers", spawn_stage(self, Default::default()));
     }
 
+    fn reduce_block(&mut self, block: &MultiEraBlock) -> Result<(), gasket::error::Error> {
+        self.output.send(gasket::messaging::Message::from(
+            CRDTCommand::block_starting(block),
+        ))?;
+
+        for reducer in self.reducers.iter_mut() {
```

Comment:
Is this where we plan to use rayon or something to do these in parallel?

Reply:
yes, exactly